crates/datasources/src/bson/mod.rs

```diff
@@ -0,0 +1,281 @@
+use datafusion::arrow::array::cast::as_string_array;
+use datafusion::arrow::array::{types::*, Array, ArrayRef, AsArray, StructArray};
+use datafusion::arrow::datatypes::{DataType, Fields, IntervalUnit, TimeUnit};
+use datafusion::arrow::error::ArrowError;
+
+use bson;
+
+pub struct BsonBatchConverter {
+    batch: StructArray,
+    schema: Vec<String>,
+    row: usize,
+    started: bool,
+    columns: Vec<Vec<bson::Bson>>,
+}
+
+impl BsonBatchConverter {
+    pub fn new(batch: StructArray, fields: Fields) -> Self {
+        let mut field_names = Vec::<String>::with_capacity(fields.len());
+        for field in &fields {
+            field_names.push(field.name().to_owned())
+        }
+
+        Self {
+            batch: batch.clone(),
+            schema: field_names,
+            row: 0,
+            started: false,
+            columns: Vec::<Vec<bson::Bson>>::with_capacity(batch.num_columns()),
+        }
+    }
+
+    fn setup(&mut self) -> Result<(), ArrowError> {
+        for col in self.batch.columns().iter() {
+            self.columns
+                .push(array_to_bson(col).map_err(|e| ArrowError::from_external_error(Box::new(e)))?)
+        }
+        self.started = true;
+        Ok(())
+    }
+}
+
+impl Iterator for BsonBatchConverter {
+    type Item = bson::Document;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if !self.started {
+            self.setup().ok()?;
+        }
```

Comment:
We lose the error here and the iterator just stops. If we can't convert something to bson, there's no indication to the user that we failed to do so, and instead write out partial data.

Would it make sense to have the iterator return `Result<bson::Document, ArrowError>`?